### DISCUSSION TITLE
Missing Exif metadata when `vips copy` to JXL

### DISCUSSION THREAD

**Discussion by cmahnke** [libvips/libvips#4173] [Help wanted]
I'm not sure if this is a missing feature, a bug or something else:
When converting a TIFF file to JXL using `vips copy` the Exif metadata is missing, in general Exif in JXL is a problem, the encoder provided by LibJXL, `cjxl` has it's problem as well. I also tried to use `exiv2` and `exiftool` but had no luck. Theew are indications on the web that the later might work, but didn't manage to get it to work with an image created by `vips copy` - maybe there is a required container missing?

How are other doing this? 

# Edit

It seems that `exiftool` installed via `brew` doesn't bundle `IO::Compress::Brotli` and `IO::Uncompress::Brotli` required for JXL, see https://github.com/exiftool/exiftool/blob/master/README 
I've created an issue: https://github.com/Homebrew/homebrew-core/issues/192273

But am still interested in your workflows / ideas

**Comment by jcupitt** [comment]
Hello @cmahnke,

TIFF has its own strange way of handling exif (it appears as a TIFF subdirectory) and libvips doesn't support this. We've talked a few times about adding EXIF support to the tiff loader, but it's not straightforward and no one's done it. The libvips TIFF loader does support XMP, IPTC, ICC etc. and they should all work with JXL.

libvips 8.16, due soon, supports EXIF in JXL. For example with this test image:

```
$ vipsheader -a k2.jpg
k2.jpg: 1450x2048 uchar, 3 bands, srgb, jpegload
width: 1450
height: 2048
bands: 3
format: uchar
coding: none
interpretation: srgb
xoffset: 0
yoffset: 0
xres: 2.835
yres: 2.835
filename: k2.jpg
vips-loader: jpegload
jpeg-multiscan: 0
jpeg-chroma-subsample: 4:2:0
exif-data: 186 bytes of binary data
resolution-unit: in
exif-ifd0-Orientation: 1 (Top-left, Short, 1 components, 2 bytes)
exif-ifd0-XResolution: 72009/1000 (72.009, Rational, 1 components, 8 bytes)
exif-ifd0-YResolution: 72009/1000 (72.009, Rational, 1 components, 8 bytes)
exif-ifd0-ResolutionUnit: 2 (Inch, Short, 1 components, 2 bytes)
exif-ifd0-YCbCrPositioning: 1 (Centred, Short, 1 components, 2 bytes)
exif-ifd2-ExifVersion: Exif Version 2.1 (Exif Version 2.1, Undefined, 4 components, 4 bytes)
exif-ifd2-ComponentsConfiguration: Y Cb Cr - (Y Cb Cr -, Undefined, 4 components, 4 bytes)
exif-ifd2-FlashpixVersion: FlashPix Version 1. (FlashPix Version 1., Undefined, 4 components, 4 bytes)
exif-ifd2-ColorSpace: 65535 (65535, Short, 1 components, 2 bytes)
exif-ifd2-PixelXDimension: 1450 (1450, Long, 1 components, 4 bytes)
exif-ifd2-PixelYDimension: 2048 (2048, Long, 1 components, 4 bytes)
orientation: 1
```

I see:

```
$ vips copy k2.jpg x.jxl
$ vipsheader -a memory: high-water mark 11.90 MB
x.jxl: 1450x2048 uchar, 3 bands, srgb, jxlload
width: 1450
height: 2048
bands: 3
format: uchar
coding: none
interpretation: srgb
xoffset: 0
yoffset: 0
xres: 2.835
yres: 2.835
filename: x.jxl
vips-loader: jxlload
icc-profile-data: 536 bytes of binary data
exif-data: 186 bytes of binary data
resolution-unit: in
exif-ifd0-Orientation: 1 (Top-left, Short, 1 components, 2 bytes)
exif-ifd0-XResolution: 72009/1000 (72.009, Rational, 1 components, 8 bytes)
exif-ifd0-YResolution: 72009/1000 (72.009, Rational, 1 components, 8 bytes)
exif-ifd0-ResolutionUnit: 2 (Inch, Short, 1 components, 2 bytes)
exif-ifd0-YCbCrPositioning: 1 (Centred, Short, 1 components, 2 bytes)
exif-ifd2-ExifVersion: Exif Version 2.1 (Exif Version 2.1, Undefined, 4 components, 4 bytes)
exif-ifd2-ComponentsConfiguration: Y Cb Cr - (Y Cb Cr -, Undefined, 4 components, 4 bytes)
exif-ifd2-FlashpixVersion: FlashPix Version 1. (FlashPix Version 1., Undefined, 4 components, 4 bytes)
exif-ifd2-ColorSpace: 65535 (65535, Short, 1 components, 2 bytes)
exif-ifd2-PixelXDimension: 1450 (1450, Long, 1 components, 4 bytes)
exif-ifd2-PixelYDimension: 2048 (2048, Long, 1 components, 4 bytes)
orientation: 1
bits-per-sample: 8
```

**Reply by cmahnke** [reply]
Thanks, I guess I'm not really stuck with Exif. Sure the scanner provides this (no XMP data is added) but it should be possible to add some translation step into the processing workflow before passing it to `vips`, but I need to check if tool like `jxlpy` can handle them.

I also figured out how to solve the container problem with `exiftool`, one needs to add the `-m` option to create missing containers.

**Reply by Qhilm** [reply]
Just a clarification, `-m` is for `--ignoreMinorErrors` not really for creating missing containers.

**Reply by cmahnke** [reply]
Thanks @Qhilm 
But it works as expected anyways
